### ISSUE TITLE
Add "sign-in-sheet handling page"

### ISSUE THREAD

**Issue by cecilia-donnelly** [OpenTechStrategies/swoptact#13]
Most data entry into this system will occur because a user has a paper sign-in sheet from an Action and needs to enter the Action information and the names of attendees into Swoptact.  This page will make that a seamless process.

At top of page, search/create Action information: name, date, site.  If the Action already exists (if this is the second or nth sign-in sheet from the Action, or if the Action was created beforehand) then users will be able to start typing its name and choose it from the filtered completion options (just as they should be able to do for people, in the next paragraph).  If it does not exist, then entering the information here will auto-save it.

Below the Action information, show a table with columns name, institution, phone (these three are the most important, and must appear first), and other participant information so a user can start typing a name and see options below (e.g., typing “Ma” shows “Maria, Mary, Matthew” below, where Maria, Mary, and Matthew are previously entered people in the database). When the user chooses one of these previously entered people, his/her information fills in the remaining columns. If the Participant being entered is a new Participant, then that new record is saved in the db, _and_ that new Participant record is linked to this Action.  When a row is entered, a new empty one appears below, and the previous entries stay on the screen.  Each row (each Participant linked to the Action) should auto-save, in case something goes wrong before they finish entering the whole sign-in sheet.

**Comment by cecilia-donnelly** [comment]
As I understand it, this issue requires:

1. A new template to show editable information in a table format.
2. An add-on to save information automatically (c.f. https://github.com/theatlantic/django-autosave), or just some javascript to save information onchange.
3. Some ability to both search and save in the same field.  We need this in several places, but I'm not sure what to call it exactly.
4. A template that combines event information with person information -- maybe duplicative of number 1.

**Comment by cecilia-donnelly** [comment]
@xray7224 and I will be working together on this, following these steps:

* add a basic custom view which registers a new URL
* create a template
* server the template ( @xray7224, do you want to clarify on this piece?  I'm going from our IRC chat)
* on POST process the form and save it
* look up all the people and events as efficiently as possible and put them in the page so Javascript can use them as suggestions
* Show autofill suggestions via Javascript, and fill remaining fields with information from the chosen object

**Comment by cecilia-donnelly** [comment]
@xray7224:
I *think* that maybe what we want to do here is actually change the "event change" page to use inline admin for participants.  That is, the sign in sheet page _will_ be the event page.  Does that seem reasonable to you?  The only trouble is that I can't seem to figure out how to get the participant fields to show up using the inline form.  It currently shows the model that _links_ events and participants, so doesn't have any of the participant fields, just the participant ID (represented in the UI by a dropdown of participants).  Since we want to be able to edit and add participants all on the same page, we need to use the _participant_ model here, not the event-participant relationship model.  Thoughts on how to do that?

This would be pretty easy using SQL joins, but I'm not sure how to get there using Django.  Basically, instead of Events.participants.through as the model we need to use Events.participants.through.participants, so we get the set of participants that are linked to this event, but that isn't an option.

**Comment by cwebber** [comment]
I originally wrote this on #14, but it's more appropriate here, so moving it.

@cecilia-donnelly asked me to look into why the inline representation of the many-to-many display of "participants" wasn't showing up as anything other than a dropdown.  Django's docs [appear to give the answer for this](https://docs.djangoproject.com/en/1.7/ref/contrib/admin/#working-with-many-to-many-models) but in reality it seems that Cecilia has correctly executed the advice from this and given [other conversations on the internet](http://stackoverflow.com/questions/10904848/adding-inline-many-to-many-objects-in-django-admin), very confusingly seeing the many to many dropdown is "expected"??? Which doesn't seem right.

Suggestions on that thread are that we subclass the form and modify the [.form attribute](https://docs.djangoproject.com/en/1.7/ref/contrib/admin/#django.contrib.admin.ModelAdmin.form) in order to get the behavior we want.  Which, okay, though that seems like a surprising amount of work for something that seemed implied like it would work having followed the docs' instructions for many2many fields.

**Comment by cwebber** [comment]
Okay, I did some research into this (thanks to people in the #django channel on freenode for helping me realize what was going on).  The behavior of the dropdown on the inline form is the expected behavior, which is strange because I'm not not sure why it would be documented the way it is without making it clear that it isn't particularly helpful behavior, but I suppose isn't surprising becuase ".through" *is* the intermediate table.

What I didn't understand until now is *why* there is no builtin way of inlining the many to many table so that the fields on the other side show up.  I now understand why this is.  If it'd be done, it would not be as simple looking as the simple foreign key inline setup, because that is a "parent->child" kind of setup... you're just adding new attached children objects inline to the main table for that page, essentially.  Which, okay, that makes sense!  You are just either adding new ones or removing them, and that's it.

But things aren't so simple once you get to a many to many setup.  In such a setup in addition to adding entirely new objects or removing them entirely, you might actually pick an existing object.  In that case, you now need to load all the fields from that object after you pick it... there's only two ways to do that: you either have to do an AJAX'y type call to get the data after the user selects it and load it in, or you have to preload and serialize *all data* from *all objects* onto the page.  The latter is probably not usually a safe or advisable thing to automatically enable in a system like Django's admin, which tries to be pretty general.  And as for the former, Django's admin doesn't do anything Ajax'y out of the box.

So what does that mean?  Well we will probably either need to add our own widget or define our own view.  Now the original description for this bug, before we started looking into using Django's magical inline behavior, suggested an AJAX'y type workflow already.  We probably want to do something along those lines.  (I have a suggested adjustment to it, but I'll make it after this braindump.)  One way or another, we'll probably need to set up some new endpoints for fetching, updating, removing connections to such many to many objects.

Maybe we can generalize our system for ajax'y inline widget + endpoints into something reusable.

I hope that clarifies why this was so confusing!

**Comment by cwebber** [comment]
I think it's probably best to wait to do further work on this one until @cecilia-donnelly and I talk.

**Comment by cwebber** [comment]
The initial version we work on will not do "autosave" or "autolookup while typing" because that's harder to code and easier to make serious mistakes in (and provide a UI with unintended effects), but may be added later.  For first round, we've agreed upon a UI that looks like this:

```
| Stuff   [_43___]                                                |
| Label   [_woop_boop__________]                                  |
|                                                                 |
|                                                      [ Save ]   |
'-----------------------------------------------------------------' 

.-----------------------------------------------------------------.
| Participants (inline)                                           |
|  *Name*     *Institution*      *Phone*                          |
| Sarah      Mtgomery High     555 614 2273   [Edit ✎] [Unlink ✘] |
| James      Camp Cool         555 819 9561   [Edit ✎] [Unlink ✘] |
| [Jill___]  [Mt_Awesome_|V]   [555_819_9561] [Save ✎] [Cancel ✖] | <- editing
| Fred       Flintrock Qry     555 129 9156   [Edit ✎] [Unlink ✘] |
|                                                                 |
| [Add new]   [Link existing]                                     |
'-----------------------------------------------------------------'
```

**Comment by cwebber** [comment]
I spent some time researching how to preserve the page but get a custom widget where the current inline widget is sitting.  It took a bunch of reading of code to find out that it's not too hard to make use of what already exists, since it turns out there's a `template=` attribute right on the `InlineModelAdmin` class.  The answer was right in front of us!

As such, I modified this to switch out the template class.  Before this can be used in an ajax'y way, I still need to figure out how to embed the model ID and javascript.  I will try to stub out as much as I can.

**Comment by cecilia-donnelly** [comment]
The current plan, constructed with @cwebber and @xray7224:

# Actions from the viewable file:
- get a list of all current many to many relationships so the initial form can be populated
- click "edit" to switch out the text with textboxes and etc
- save changes to return a success response, 400 error (with json strings) or a 500 errror
- Unlink a participant from the current event
- Add an existing participant
- Create a participant and add to this event (luckily, this can use the endpoint for save changes, depending on whether an id [for update existing] is sent or not [for create])


# Endpoints:
All endpoints will return HTTP status code 200 for successful requests,  HTTP status code 400 for errors that occurred processing the request,  HTTP status code 401 for not authorized, and 500 for server error 500 (usual django behavior).

- GET /api/event/<event id>/participants/
    This lists all the participants in a JSON list that are "linked" to the event
- GET /api/participant/<participant_id>/
    This gets all the information about the participant with a given ID
- PUT /api/participants/<participant_id>/
    This is to update a existing participant. The body of this request should be a JSON
    serialized copy of the participant which you would like to save.
- POST /api/participants/
    This is to create new participants. The body of the request should be a JSON serialized
    participant.
- DELETE /api/events/<event_id>/participants/<participant_id>/
    This will "unlink" the participant from the event. This will NOT delete the participant.
- GET /api/events/<event_id>/available-participants/
    This lists all the participants in a JSON list that are NOT linked to the event.
- POST /api/events/<event_id>/participants/<participant_id>/
    This will link the person with with the given participant ID to the event. Any body provided
    will be ignored.

The javascript handling each ajax call should have a success callback as well as a failure callback; in case of failure, 400 vs 500 will both be handled, but the errors from a 400 will be parsed as json (500, could just say "server error")


# TODO:
@cwebber:
* Fix "fail to save" issue
* Add "commented out mockups" of what examples of a currently listed participant vs edit-in-progress are, an edit in progress with errors, etc (cecilia can reproduce these by manipulating the DOM, cwebber should add notes/examples)
* Write notes in js file for Cecilia? (yes please)

Cecilia:
* Write javascript to offer both display and textboxes (for editing) of participants
 * Send info Jess's endpoints

@xray7224:
* write endpoints described above

**Comment by cwebber** [comment]
"Fix "fail to save" issue": done, by actually removing the widget altogether and making a custom template.
In retrospect, the obvious and right way to do it, but it took a long time of staring at django internals to realize this.

**Comment by cwebber** [comment]
I finished putting all the "mockup elements" in place: http://dustycloud.org/tmp/swoptact_mockup_fields.png

**Comment by cecilia-donnelly** [comment]
Remaining bugs:
* edit button doesn't trigger correctly after editing and saving of its row (that is, if I click "edit" then "save" then "edit" again, it won't do anything the second time)
* Institution is not editable from sign in sheet
* Person missing an institution is not editable (error: TypeError at /api/participants/36/argument of type 'NoneType' is not iterable).  I'm not sure where the institution needs to be iterable.  Maybe @xray7224 knows?
* Address is not editable (I'm making it a straight charField for the demo tomorrow, to get around this.  Eventually we'll want to make it possible to enter an address into a text field and parse it into its constituent parts).

**Comment by xray7224** [comment]
Yeah it's coming from https://github.com/OpenTechStrategies/swoptact/blob/master/swoptact/views.py#L69. The bug is easily fixed prior to this line do something like:
```
if value is None:
    continue
```
If the bug still exists when i do my OTS work i shall fix it.

**Comment by cecilia-donnelly** [comment]
@xray7224 thank you!  Much simpler than I feared.

**Comment by cwebber** [comment]
@cecilia-donnelly I worked > 20 hours last week getting this stuff together, so if it's okay, I'm happy to help with these bugs and any others next week instead.  Is that okay?  Or is there urgency remaining in this week to wrap this up, even though the demo is done?

**Comment by cecilia-donnelly** [comment]
@cwebber yes, I figured you'd be out for the rest of the week.  Thanks for checking in!  I think Jess and I should be able to handle these bugs this week.

**Comment by cecilia-donnelly** [comment]
User feedback on this was in the form of a UI wishlist:
- freeze top row
- rename "unlink" to "undo"
- test ESC key, etc and test on mobile devices
- fix spacing/presentation for phone numbers
- capitalize all header words
- alter entries such that address (and potentially other fields, to be explained later) will optionally show as a second row (that is, users can click on something and the second row will show up, but by default we see only name, number, and institution)
- make data entry typeface heavier and larger so that it's easier to read
- make action buttons grayscale.  Currently they scream too much.

**Comment by cecilia-donnelly** [comment]
Thoughts about the sticky table header: 
* I tried this https://github.com/markmalek/Fixed-Header-Table but it didn't work with our dynamically created table
* Useful, if complex: http://tympanus.net/codrops/2014/01/09/sticky-table-headers-columns/  Relies on the throttle+debounce plugin (?)

**Comment by cecilia-donnelly** [comment]
Current to-dos:
- [ ] Add separate address row
- [ ] Freeze header row
- [ ] Prevent save of full event while still editing a person
- [ ] replace "link existing" with autocompletion in the "add new" row
- [ ] make "add new" work multiple times
 - [ ] make "save" work multiple times

**Comment by cecilia-donnelly** [comment]
This issue was split into issues #45, #46, #47, #48, #49, #50, #51, #52.
